src/coreWebVitals.ts

```diff
@@ -0,0 +1,121 @@
+	browser_id: null,
+	page_view_id: null,
+	fid: null,
+	cls: null,
+	lcp: null,
+	fcp: null,
+	ttfb: null,
+};
+
+// By default, sample 1% of page views
+const userInSample = Math.random() < 1 / 100;
+
+let shouldSendMetrics = window.location.hash === '#forceSendMetrics';
+// unless we are forcing sending metrics for this page view
+// via initialisation or calling forceSendMetrics()
+export const forceSendMetrics = (): void => {
+	shouldSendMetrics = true;
+};
+
+const roundWithDecimals = (value: number, precision = 6): number => {
+	const power = Math.pow(10, precision);
+	return Math.round(value * power) / power;
+};
+
+const sendData = (): boolean => {
+	if (
+		!(userInSample || shouldSendMetrics) ||
+		coreWebVitalsPayload.fcp === null
+	)
+		return false;
+
+	const endpoint =
+		window.location.hostname === 'm.code.dev-theguardian.com' ||
+		window.location.hostname === 'localhost' ||
+		window.location.hostname === 'preview.gutools.co.uk'
+			? 'https://performance-events.code.dev-guardianapis.com/core-web-vitals'
+			: 'https://performance-events.guardianapis.com/core-web-vitals';
+
+	return navigator.sendBeacon(endpoint, JSON.stringify(coreWebVitalsPayload));
+};
+
+const onReport: ReportHandler = (metric) => {
+	switch (metric.name) {
+		case 'FCP':
+			// Browser support: Chromium, Firefox, Safari Technology Preview
+			coreWebVitalsPayload.fcp = roundWithDecimals(metric.value);
+			break;
+		case 'CLS':
+			// Browser support: Chromium,
+			coreWebVitalsPayload.cls = roundWithDecimals(metric.value);
+			break;
+		case 'LCP':
+			// Browser support: Chromium
+			coreWebVitalsPayload.lcp = roundWithDecimals(metric.value);
+			break;
+		case 'FID':
+			// Browser support: Chromium, Firefox, Safari, Internet Explorer (with the polyfill)
+			coreWebVitalsPayload.fid = roundWithDecimals(metric.value);
+			break;
+		case 'TTFB':
+			// Browser support: Chromium, Firefox, Safari, Internet Explorer
+			coreWebVitalsPayload.ttfb = roundWithDecimals(metric.value);
+			break;
+	}
+};
+
+export const coreWebVitals = (
+	{
+		browserId,
+		pageViewId,
+		forceSendMetrics = false,
+	}: {
+		browserId: string;
+		pageViewId: string;
+		forceSendMetrics: boolean;
+	},
+	metricsSentCallback: (queued?: boolean) => void = () => void 0,
+): void => {
+	coreWebVitalsPayload.browser_id = browserId;
+	coreWebVitalsPayload.page_view_id = pageViewId;
+
+	if (forceSendMetrics) shouldSendMetrics = true;
+
+	getCLS(onReport, false);
+	getFID(onReport);
+	getLCP(onReport);
+	getFCP(onReport);
+	getTTFB(onReport);
+
+	// Report all available metrics when the page is backgrounded or unloaded.
+	addEventListener('visibilitychange', () => {
+		if (document.visibilityState === 'hidden') {
+			const queued = sendData();
+			metricsSentCallback(queued);
+		}
+	});
+	// Safari does not reliably fire the `visibilitychange` on page unload.
+	addEventListener('pagehide', () => {
+		const queued = sendData();
+		metricsSentCallback(queued);
+	});
+};
```

Comment:
This approach seems like a big improvement - sending a payload which is as populated as possible once per page rather than sending cumulatively populated payload many times.
Looks like it's also the recommended way to send analytics data https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event#usage_notes

Reply:
Yes, I was going to add a note about this.

It’s also recommended by `web-vitals` https://github.com/GoogleChrome/web-vitals#batch-multiple-reports-together